api/v1alpha1/stage_types.go

```diff
@@ -185,10 +185,25 @@ func (s *Stage) IsFreightAvailable(freight *Freight) bool {
 			if req.Sources.Direct {
 				return true
 			}
-			for _, source := range req.Sources.Stages {
-				if freight.IsVerifiedIn(source) {
-					return req.Sources.RequiredSoakTime == nil ||
-						freight.GetLongestSoak(source) >= req.Sources.RequiredSoakTime.Duration
+			if req.Sources.AvailabilityStrategy == FreightAvailabilityStrategyAll {
+				// Make sure Freight is verified and soaked in all upstream Stages
+				for _, stage := range req.Sources.Stages {
+					_, verified := freight.Status.VerifiedIn[stage]
+					soaked := req.Sources.RequiredSoakTime == nil ||
+						freight.GetLongestSoak(stage) >= req.Sources.RequiredSoakTime.Duration
+					if !verified || !soaked {
+						return false
+					}
+				}
+				return true
+			}
+			// Make sure Freight is verified and soaked in at least one upstream Stage
+			for _, stage := range req.Sources.Stages {
+				_, verified := freight.Status.VerifiedIn[stage]
+				soaked := req.Sources.RequiredSoakTime == nil ||
+					freight.GetLongestSoak(stage) >= req.Sources.RequiredSoakTime.Duration
+				if verified && soaked {
+					return true
```

Comment:
It might be nice to make this logic a method on `Freight`, so that we can avoid the logic duplication in this and the above loop.

Reply:
That had occurred to me at one point and I think I didn't entertain it very long because in the other spot, a lot of the work is done through field selectors, but this aspect of it isn't, so I shouldn't have dismissed it so quickly. 👀

Reply:
Oh... sorry... I misunderstood. You mean the loop immediately above this. Yes. But I think there's even more opportunity than that to DRY things up.